plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,545 @@
+				g_debug ("fail to open %s", remove_path);
+			}
+
+			if (f) {
+				value = '1';
+				if (fwrite (&value, 1, 1, f) != 1) {
+					g_debug ("fail to write %s", remove_path);
+				}
+			}
+
+			if (f)
+				fclose(f);
+
+			argc = g_new0 (gchar *, 4);
+			argc[3] = NULL;
+			argc[0] = g_strdup ("killall");
+			argc[1] = g_strdup ("-9");
+			argc[2] = g_strdup ("mbim-proxy");
+			if (!g_spawn_sync (NULL,
+					  argc,
+					  NULL,
+					  G_SPAWN_SEARCH_PATH | G_SPAWN_STDOUT_TO_DEV_NULL | G_SPAWN_STDERR_TO_DEV_NULL,
+					  NULL,
+					  NULL,
+					  NULL,
+					  NULL,
+					  &exit_stat,
+					  &tmp_error)) {
+				g_debug ("error spawning kill mbim-proxy: %s ", tmp_error->message);
+				g_clear_error (&tmp_error);
+			}
+
+			if (!(f = fopen (rescan_path, "w"))) {
+				g_debug ("fail to open %s", rescan_path);
+			}
+
+			if (f) {
+				value = '1';
+				if (fwrite (&value, 1, 1, f) != 1) {
+					g_debug ("fail to write %s", rescan_path);
+				}
+			}
+
+			if (f)
+				fclose (f);
+			g_debug ("%s", remove_path);
+			g_debug ("%s", rescan_path);
+		} else
+			return FALSE;
+	}
+
+	g_warn_if_fail (ctx.timeout_id == 0);
+	ctx.timeout_id = g_timeout_add_seconds (15, fu_mbim_qdu_updater_reprobe_timeout, &ctx);
+	g_main_loop_run (mainloop);
+
+	//update detach right after this
+	return TRUE;
+}
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+  	GCancellable    *cancellable;
+	GError		*error;
+	guint		 timeout_id;
+	GBytes		*blob;
+	GArray		*digest;
+	GPtrArray	*chunks;
+	guint 		chunk_sent;
+	FuDevice 	*device;
+} WriteContext;
+
+static gboolean
+fu_mbim_qdu_updater_reboot_timeout (gpointer user_data)
+{
+	WriteContext *ctx = user_data;
+
+	ctx->timeout_id = 0;
+	g_main_loop_quit (ctx->mainloop);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fu_mbim_qdu_updater_file_write_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	MbimMessage *request  = NULL;
+	MbimMessage *response = NULL;
+	g_autoptr(GError)      error = NULL;
+	WriteContext *ctx = user_data;
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_debug ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_file_write_response_parse (
+							 response,
+							 &error)) {
+		g_debug ("[%s][%d]: error: couldn't parse response message: %s\n", __FUNCTION__, __LINE__, error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	fu_device_set_progress_full (FU_DEVICE (ctx->device), (gsize) ctx->chunk_sent, (gsize) ctx->chunks->len);
+	ctx->chunk_sent++;
+	if (ctx->chunk_sent < ctx->chunks->len) {
+		FuChunk *chk = g_ptr_array_index (ctx->chunks, ctx->chunk_sent);
+		request = (mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk), (const guint8 *)fu_chunk_get_data (chk), &error));
```

Comment:
Pass NULL at the GError, and assume set_new() never fails. Also, maybe drop the extra () around the method call